File: DesignPatterns/2-strategyPattern/cpp/main.cpp
/*
 * File: main.cpp
 * Created Date: 2021-12-12 05:06:54
 * Author: ysj
 * Description:  cpp策略模式
 */
#include <iostream>
#include "strategy.hpp"
#include "context.hpp"
using namespace std;

int main()
{
    // 购买价格、数量
    float price = 599;
    float quantity = 1;
    float originPrice = price * quantity;
    cout << "单价:" << price << " 数量:" << quantity << " 原价:" << originPrice << endl;

    // 正常计算策略
    PriceCalculationStrategy *keepNormal = new KeepNormalStrategy();
    PriceCalculationContext *ctx = new PriceCalculationContext(keepNormal);
    float totalPrice = ctx->calculatePrice(originPrice);
    cout << "正常价格:" << totalPrice << endl;

    // 打折计算策略
    PriceCalculationStrategy *discount = new DiscountStrategy(0.8);
    ctx = new PriceCalculationContext(discount);
    totalPrice = ctx->calculatePrice(originPrice);
    cout << "八折价格:" << totalPrice << endl;

    // 满减策略
    PriceCalculationStrategy *fullReduction = new FullReductionStrategy(500, 200);
    ctx = new PriceCalculationContext(fullReduction);
    totalPrice = ctx->calculatePrice(originPrice);
    cout << "满500减200价格:" << totalPrice << endl;

    // 无门槛策略
    PriceCalculationStrategy *directReduction = new DirectReductionStrategy(100);
    ctx = new PriceCalculationContext(directReduction);
    totalPrice = ctx->calculatePrice(originPrice);
    cout << "直减100价格:" << totalPrice << endl;
    
    return 0;
}

File: DesignPatterns/2-strategyPattern/golang/main.go
/*
 * File: main.go
 * Created Date: 2021-12-12 04:26:03
 * Author: ysj
 * Description:  golang策略模式
 */

package main

import "fmt"

func main() {
	// 购买价格、数量
	price := 599.0
	quantity := 1.0
	originPrice := price * quantity
	fmt.Printf("单价:%.2f 数量:%.2f 原价:%.2f\n", price, quantity, originPrice)

	// 正常计算策略
	keepNormal := NewKeepNormalStrategy()
	ctx := NewPriceCalculationContext(keepNormal)
	totalPrice := ctx.calculatePrice(originPrice)
	fmt.Printf("正常价格:%.2f\n", totalPrice)

	// 打折计算策略
	discount := NewDiscountStrategy(0.8)
	ctx = NewPriceCalculationContext(discount)
	totalPrice = ctx.calculatePrice(originPrice)
	fmt.Printf("八折价格:%.2f\n", totalPrice)

	// 满减策略
	fullReduction := NewFullReductionStrategy(500, 200)
	ctx = NewPriceCalculationContext(fullReduction)
	totalPrice = ctx.calculatePrice(originPrice)
	fmt.Printf("满500减200价格:%.2f\n", totalPrice)

	// 无门槛策略
	directReduction := NewDirectReductionStrategy(100)
	ctx = NewPriceCalculationContext(directReduction)
	totalPrice = ctx.calculatePrice(originPrice)
	fmt.Printf("直减100价格:%.2f\n", totalPrice)
}
